File: 0x08-python-more_classes/4-rectangle.py
#!/usr/bin/python3

"""Class Rectangle definition"""


class Rectangle:

    """Model of a Rectangle"""

    def __init__(self, width=0, height=0):

        """__init__ method for the Rectangle class
        Arguments:
        private instance width: integer that defines the width of rectangle
        private instance height: integer that defines the height of rectangle
        """

        self.width = width
        self.height = height

    """property to retrieve width of the rectangle"""

    @property
    def width(self):
        return self.__width

    """property to set the value of width of the rectangle"""

    @width.setter
    def width(self, value):
        if type(value) is not int:
            raise TypeError("width must be an integer")
        if value < 0:
            raise ValueError("width must be >= 0")

        self.__width = value

    """property to retreive height of the rectangle"""

    @property
    def height(self):
        return self.__height

    """property to set the value of height of rectangle"""

    @height.setter
    def height(self, value):
        if type(value) is not int:
            raise TypeError("height must be an integer")
        if value < 0:
            raise ValueError("height must be >= 0")

        self.__height = value

    """public instance method to return rectangle area"""

    def area(self):
        return self.__width * self.__height

    """public instance method to return rectangle perimeter"""

    def perimeter(self):
        if self.__width == 0 or self.__height == 0:
            return 0
        return (2 * self.__width + 2 * self.__height)

    """str method to print a given rectangle"""

    def __str__(self):
        result = ""
        if self.__width == 0 or self.__height == 0:
            return result
        else:
            for element in range(self.__height):
                result += "#" * self.__width
                if element != self.__height - 1:
                    result += "\n"
        return result

    def __repr__(self):
        return "Rectangle( {}, {})".format(self.__width, self.__height)

File: 0x13-javascript_objects_scopes_closures/4-rectangle.js
#!/usr/bin/node

class Rectangle {
  constructor (w, h) {
    if (w > 0 && h > 0) {
      this.width = w;
      this.height = h;
    }
  }

  print () {
    let string = '';
    for (let column = 0; column < this.height; column++) {
      for (let row = 0; row < this.width; row++) {
        string += 'X';
      }
      if (column !== this.height - 1) {
        string += '\n';
      }
    }
    console.log(string);
  }

  rotate () {
    const tmp = this.width;
    this.width = this.height;
    this.height = tmp;
  }

  double () {
    this.width = this.width * 2;
    this.height = this.height * 2;
  }
}

module.exports = Rectangle;
